### PR TITLE
Fix inconsistencies in mount/umount parameters.

### PR DESCRIPTION
**Pull request by luis4a0**
Path names are now normalized before mounting or unmounting.

This way, directories can be mounted using relative paths and unmounted using absolute paths or viceversa. Also, the final slash in the path can be omitted.

The fix works in the multipass client, so the daemon always receives an absolute mountpoint as parameter.

---

Actual diff: https://github.com/canonical/multipass/compare/fix-mount-dir-permissions...fix-mount-umount-inconsistencies

### CODE REVIEW

File: src/daemon/daemon.cpp [hunk 1]
```diff
@@ -1241,9 +1241,40 @@ try // clang-format on
         }
 
         auto& vm = it->second;
+        auto& vm_specs = vm_instance_specs[name];
 
         if (vm->current_state() == mp::VirtualMachine::State::running)
         {
+            // If the given path is relative, we need to resolve the fill
```
Comment (ricab) [inline src/daemon/daemon.cpp]:
```suggestion
            // If the given path is relative, we need to resolve the full
```

File: src/daemon/daemon.cpp [hunk 2]
```diff
@@ -1241,9 +1241,40 @@ try // clang-format on
...
+
+                target_path = mp::utils::normalize_absolute_path(home + '/' + target_path);
+                mpl::log(mpl::Level::info, category, fmt::format("home directory: {}, target_path: {}", home, target_path));
+            }
+            else
+            {
+                target_path = mp::utils::normalize_absolute_path(target_path);
+                mpl::log(mpl::Level::info, category, fmt::format("target_path: {}", target_path));
+            }
+
```
Comment (ricab) [inline src/daemon/daemon.cpp]: Since this code is repeated below, could it be extracted into a helper function, to avoid duplication? Something returning the resolved and normalized `target_path` from the given the VM name and `target_path` input.

File: src/sshfs_mount/sshfs_mount.cpp [hunk 3]
```diff
@@ -59,30 +60,56 @@ void check_sshfs_exists(mp::SSHSession& session)
     run_cmd(session, "which sshfs", error_handler);
 }
 
-void make_target_dir(mp::SSHSession& session, const std::string& target)
+// Split a path into existing and to-be-created parts. The input is assumed to
+// be an absolute path, normalized with mp::utils::normalize_absolute_path().
+std::pair<std::string, std::string> get_path_split(mp::SSHSession& session, const std::string& target)
```
Comment (ricab) [inline src/sshfs_mount/sshfs_mount.cpp]: Isn't this part of the other PR? In such occasions, where you need the changes of a PR that still hasn't gone in, it is better to stack the PRs. So you'd branch this PR out of where the other PR left off.

File: src/utils/utils.cpp [hunk 4]
```diff
@@ -342,3 +342,10 @@ void mp::utils::check_and_create_config_file(const QString& config_file_path)
         config_file.open(QIODevice::WriteOnly);
     }
 }
+
+// Given an absolute path, removes trailing slash, multiple slashes and
+// redundant "." and "..".
+std::string mp::utils::normalize_absolute_path(const std::string& path)
+{
+    return QDir::cleanPath(QString::fromStdString(path)).toStdString();
```
Comment (ricab) [inline src/utils/utils.cpp]: Nice, `QDir::cleanPath` indeed seems the right choice for this. Symlinks are left alone, but that is probably ok, and if we decide we should resolve them, that can always done in a separate PR.

Comment (luis4a0) [reply src/utils/utils.cpp]: Sure. Symlinks can be resolved exactly as we executed 'pwd' when mounting/unmounting. We'll see that later.

File: src/sshfs_mount/sshfs_mount.cpp (hunk 3 above)
Comment (luis4a0) [reply src/sshfs_mount/sshfs_mount.cpp]: Yes, that's part of the other PR. There can be a problem when merging we'd need to solve. That can be done. I can try to stack the PR's, would be a good idea (altough I'll need to force-push at some point).

File: src/daemon/daemon.cpp (hunk 2 above)
Comment (luis4a0) [reply src/daemon/daemon.cpp]: Nice, I'll do it.

File: src/daemon/daemon.cpp (hunk 1 above)
Comment (luis4a0) [reply src/daemon/daemon.cpp]: Fixed.

File: src/utils/utils.cpp (hunk 4 above)
Comment (ricab) [reply src/utils/utils.cpp]: Right, something along those lines, like `pwd -P` or `readlink -f`. Something to think about.

**Comment by luis4a0** [comment]
It's strange, lcov says that nothing is tested from the daemon functions. I'll take a look to the tests.

File: src/sshfs_mount/sshfs_mount.cpp (hunk 3 above)
Comment (ricab) [reply src/sshfs_mount/sshfs_mount.cpp]: Right, it is generally a good idea to avoid commit overlap. So stacking branches/PRs decouples things and is usually better (even though GH is not very good dealing with those). 

On force-pushing, I know it is generally not advised, but it is pretty much part of our usual workflow for PR branches (like when rebasing on the tip of master). It is fine as long as no one depends on the changing refs, so we usually avoid depending on each other's branches and we're ok if we don't force-push master.

Comment (luis4a0) [reply src/sshfs_mount/sshfs_mount.cpp]: Well, we can manually merge on this specific case, there will be at most one conflict and the file on this commit is what will prevail. Next time I'll pay attention to this issue.

About force pushing, I'm easy. Since I am the only one pushing to this branch, there will be no issues.

Comment (luis4a0) [reply src/sshfs_mount/sshfs_mount.cpp]: Solved, now this branch is on top of the other PR.

File: tests/test_sshfsmount.cpp [hunk 5]
```diff
@@ -241,3 +309,52 @@ TEST_F(SshfsMount, unblocks_when_sftpserver_exits)
...
+{
+    bool invoked{false};
+    auto request_exec = make_exec_that_fails_for({"pwd"}, invoked);
+    REPLACE(ssh_channel_request_exec, request_exec);
+
+    EXPECT_THROW(make_sshfsmount(), std::runtime_error);
+    EXPECT_TRUE(invoked);
+}
+
+TEST_F(SshfsMount, executes_commands)
```
Comment (townsend2010) [inline tests/test_sshfsmount.cpp]: Sorry I'm late to this party, but this test is quite problematic.  I'm working on rebasing #1328 and have added a number of new commands and removed old ones.  This totally breaks this test and since this test is very specific to commands and order, it is not trivial to update it when we make changes elsewhere.   Looking at this test, I'm not sure what this test gains us that is not covered elsewhere.  Is this test really necessary?

**Comment by townsend2010** [comment]
Sorry for the noise in my code comment.  It was directed at #1359, but having discussed this with @Saviq, I think we have a plan on accounting for unrelated sshfs-based command changes.
